File: ts/example/sort/service/TestHeapsortService.test.ts
import { Array } from '@otterop/lang/Array';
import { String } from '@otterop/lang/String';
import { TestBase } from '@otterop/test/TestBase';
import { SortService } from './SortService';
import { HeapsortService } from './HeapsortService';

export class TestHeapsortService extends TestBase {

    public normalSort() : void {
        let item : String;
        let items : Array<String> = Array.newArray(3, (null as unknown as String));
        items.set(0, String.wrap("3"));
        items.set(1, String.wrap("2"));
        items.set(2, String.wrap("1"));
        let sortService : SortService = new HeapsortService();
        let itemsSorted : Array<String> = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 3, String.wrap("itemsSorted should have size 3"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0, String.wrap("expected 1 as first element"));
        item = itemsSorted.get(1);
        this.assertTrue(item.compareTo(String.wrap("2")) == 0, String.wrap("expected 2 as second element"));
        item = itemsSorted.get(2);
        this.assertTrue(item.compareTo(String.wrap("3")) == 0, String.wrap("expected 3 as third element"));
    }

    public singleItem() : void {
        let item : String;
        let items : Array<String> = Array.newArray(1, (null as unknown as String));
        items.set(0, String.wrap("1"));
        let sortService : SortService = new HeapsortService();
        let itemsSorted : Array<String> = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 1, String.wrap("itemsSorted should have size 1"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0, String.wrap("expected 1 as first element"));
    }

    public sameItem() : void {
        let item : String;
        let items : Array<String> = Array.newArray(3, (null as unknown as String));
        items.set(0, String.wrap("1"));
        items.set(1, String.wrap("1"));
        items.set(2, String.wrap("1"));
        let sortService : SortService = new HeapsortService();
        let itemsSorted : Array<String> = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 3, String.wrap("itemsSorted should have size 3"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0, String.wrap("expected 1 as first element"));
        item = itemsSorted.get(1);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0, String.wrap("expected 1 as second element"));
        item = itemsSorted.get(2);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0, String.wrap("expected 1 as third element"));
    }
}


test('normalSort', () => {
    new TestHeapsortService().normalSort();
});

test('singleItem', () => {
    new TestHeapsortService().singleItem();
});

test('sameItem', () => {
    new TestHeapsortService().sameItem();
});

File: go/example/sort/service/testheapsortservice_test.go
package service

import (
    "testing"
    lang "github.com/otterop/otterop/go/lang"
    test "github.com/otterop/otterop/go/test"
)


func testHeapsortServiceNew() *TestHeapsortService {
    this := new(TestHeapsortService)
    this.TestBase = test.TestBaseNew()
    return this
}

type TestHeapsortService struct {
    *test.TestBase
}




func (this *TestHeapsortService) normalSort()  {
    var item *lang.String
    var items *lang.Array[*lang.String] = lang.ArrayNewArray(3, (*lang.String)(nil))
    items.Set(0, lang.StringWrap(lang.StringLiteral("3")))
    items.Set(1, lang.StringWrap(lang.StringLiteral("2")))
    items.Set(2, lang.StringWrap(lang.StringLiteral("1")))
    var sortService SortService = HeapsortServiceNew()
    var itemsSorted *lang.Array[*lang.String] = sortService.Sort(items)
    this.AssertTrue(itemsSorted.Size() == 3, lang.StringWrap(lang.StringLiteral("itemsSorted should have size 3")))
    item = itemsSorted.Get(0)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("1"))) == 0, lang.StringWrap(lang.StringLiteral("expected 1 as first element")))
    item = itemsSorted.Get(1)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("2"))) == 0, lang.StringWrap(lang.StringLiteral("expected 2 as second element")))
    item = itemsSorted.Get(2)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("3"))) == 0, lang.StringWrap(lang.StringLiteral("expected 3 as third element")))
}

func (this *TestHeapsortService) singleItem()  {
    var item *lang.String
    var items *lang.Array[*lang.String] = lang.ArrayNewArray(1, (*lang.String)(nil))
    items.Set(0, lang.StringWrap(lang.StringLiteral("1")))
    var sortService SortService = HeapsortServiceNew()
    var itemsSorted *lang.Array[*lang.String] = sortService.Sort(items)
    this.AssertTrue(itemsSorted.Size() == 1, lang.StringWrap(lang.StringLiteral("itemsSorted should have size 1")))
    item = itemsSorted.Get(0)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("1"))) == 0, lang.StringWrap(lang.StringLiteral("expected 1 as first element")))
}

func (this *TestHeapsortService) sameItem()  {
    var item *lang.String
    var items *lang.Array[*lang.String] = lang.ArrayNewArray(3, (*lang.String)(nil))
    items.Set(0, lang.StringWrap(lang.StringLiteral("1")))
    items.Set(1, lang.StringWrap(lang.StringLiteral("1")))
    items.Set(2, lang.StringWrap(lang.StringLiteral("1")))
    var sortService SortService = HeapsortServiceNew()
    var itemsSorted *lang.Array[*lang.String] = sortService.Sort(items)
    this.AssertTrue(itemsSorted.Size() == 3, lang.StringWrap(lang.StringLiteral("itemsSorted should have size 3")))
    item = itemsSorted.Get(0)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("1"))) == 0, lang.StringWrap(lang.StringLiteral("expected 1 as first element")))
    item = itemsSorted.Get(1)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("1"))) == 0, lang.StringWrap(lang.StringLiteral("expected 1 as second element")))
    item = itemsSorted.Get(2)
    this.AssertTrue(item.CompareTo(lang.StringWrap(lang.StringLiteral("1"))) == 0, lang.StringWrap(lang.StringLiteral("expected 1 as third element")))
}

func TestTestHeapsortService(t *testing.T) {
    t.Run("normalSort", func(t *testing.T) {
        test := testHeapsortServiceNew()
        test.SetGoTestingT(t)
        test.normalSort()
    })
    t.Run("singleItem", func(t *testing.T) {
        test := testHeapsortServiceNew()
        test.SetGoTestingT(t)
        test.singleItem()
    })
    t.Run("sameItem", func(t *testing.T) {
        test := testHeapsortServiceNew()
        test.SetGoTestingT(t)
        test.sameItem()
    })
}
